### ISSUE TITLE
S3 HeadObject fails with large files

### ISSUE THREAD

**Issue by jsyrjala** [cognitect-labs/aws-api#97]
## Dependencies

``` clojure
 [com.cognitect.aws/api "0.8.352"]
   [com.cognitect/http-client "0.1.99"]
     [org.eclipse.jetty/jetty-client "9.4.15.v20190215"]
       [org.eclipse.jetty/jetty-io "9.4.15.v20190215"]
     [org.eclipse.jetty/jetty-http "9.4.15.v20190215"]
     [org.eclipse.jetty/jetty-util "9.4.15.v20190215"]
   [commons-codec "1.11"]
   [org.clojure/core.async "0.4.490"]
     [org.clojure/tools.analyzer.jvm "0.7.2"]
       [org.clojure/core.memoize "0.5.9"]
       [org.clojure/tools.analyzer "0.6.9"]
       [org.clojure/tools.reader "1.0.0-beta4"]
       [org.ow2.asm/asm-all "4.2"]
   [org.clojure/data.json "0.2.6"]
   [org.clojure/data.xml "0.2.0-alpha6"]
     [org.clojure/data.codec "0.1.0"]
 [com.cognitect.aws/endpoints "1.1.11.636"]
 [com.cognitect.aws/lambda "746.2.533.0"]
 [com.cognitect.aws/s3 "747.2.533.0"]
 [nrepl "0.6.0" :exclusions [[org.clojure/clojure]]]
 [org.clojure/clojure "1.10.1"]
   [org.clojure/core.specs.alpha "0.2.44"]
   [org.clojure/spec.alpha "0.2.176"]
 [org.clojure/core.cache "0.8.2"]
   [org.clojure/data.priority-map "0.0.7"]
 [org.clojure/tools.logging "0.5.0"]
 [org.slf4j/jcl-over-slf4j "1.7.28"]
 [org.slf4j/log4j-over-slf4j "1.7.28"]
 [org.slf4j/slf4j-api "1.7.28"]
 [prismatic/schema "1.1.12"]
```

## Description with failing test case

I am trying to check the existence of a file in S3 with S3 HeadObject command. The checking fails with one file and works with another file, both files are in the same bucket. Only difference between the files I can see is their size.

The file that results in an error a big one about 15G, and the one that works is a smaller, about 400M.

The failing file
```
aws s3 ls s3://bucket/path/file.mxf
2019-09-24 18:36:56 15395867624 file.mxf
```

## Stack traces

```
(aws/invoke s3 {:op :HeadObject :request {:Bucket "bucket"
                                          :Key "path/file.mxf" }})
=>
{:cognitect.anomalies/category :cognitect.anomalies/fault,
 :cognitect.aws.client/throwable #error{:cause nil,
                                        :via [qtp1312101229-46] INFO  o.e.jetty.client.ResponseNotifier - Exception while notifying listener org.eclipse.jetty.client.HttpRequest$10@7ec5d16c
java.lang.NegativeArraySizeException: -1784001560
	at clojure.lang.Numbers.byte_array(Numbers.java:1394)
	at cognitect.http_client$empty_bbuf.invokeStatic(http_client.clj:37)
	at cognitect.http_client$on_headers.invokeStatic(http_client.clj:127)
	at cognitect.http_client$on_headers.invoke(http_client.clj:108)
	at clojure.lang.Atom.swap(Atom.java:51)
	at clojure.core$swap_BANG_.invokeStatic(core.clj:2353)
	at cognitect.http_client.Client$fn$reify__13616.onHeaders(http_client.clj:225)
	at org.eclipse.jetty.client.HttpRequest$10.onHeaders(HttpRequest.java:493)
	at org.eclipse.jetty.client.ResponseNotifier.notifyHeaders(ResponseNotifier.java:98)
	at org.eclipse.jetty.client.ResponseNotifier.notifyHeaders(ResponseNotifier.java:90)
	at org.eclipse.jetty.client.HttpReceiver.responseHeaders(HttpReceiver.java:268)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.headerComplete(HttpReceiverOverHTTP.java:256)
	at org.eclipse.jetty.http.HttpParser.parseFields(HttpParser.java:1218)
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1502)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.parse(HttpReceiverOverHTTP.java:172)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:135)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:73)
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:133)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:155)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:305)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.onFillable(SslConnection.java:427)
	at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:321)
	at org.eclipse.jetty.io.ssl.SslConnection$2.succeeded(SslConnection.java:159)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:333)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:310)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:168)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:126)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:366)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:765)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:683)
[	at java.base/java.lang.Thread.run(Thread.java:834)
{:type java.lang.NullPointerException,
                                               :message nil,
                                               :at [clojure.lang.Numbers ops "Numbers.java" 1068]}],
                                        :trace [[clojure.lang.Numbers ops "Numbers.java" 1068]
                                                [clojure.lang.Numbers lt "Numbers.java" 251]
                                                [clojure.lang.Numbers lt "Numbers.java" 3909]
                                                [cognitect.aws.protocols.rest$parse_http_response
                                                 invokeStatic
                                                 "rest.clj"
                                                 257]
                                                [cognitect.aws.protocols.rest_xml$eval13868$fn__13869
                                                 invoke
                                                 "rest_xml.clj"
                                                 21]
                                                [clojure.lang.MultiFn invoke "MultiFn.java" 239]
                                                [cognitect.aws.client$handle_http_response invokeStatic "client.clj" 47]
                                                [cognitect.aws.client$send_request$fn__13250 invoke "client.clj" 78]
                                                [clojure.core$map$fn__5862$fn__5863 invoke "core.clj" 2742]
                                                [clojure.core.async.impl.channels$chan$fn__2434
                                                 invoke
                                                 "channels.clj"
                                                 300]
                                                [clojure.core.async.impl.channels.ManyToManyChannel
                                                 put_BANG_
                                                 "channels.clj"
                                                 83]
                                                [clojure.core.async$put_BANG_ invokeStatic "async.clj" 165]
                                                [cognitect.http_client.Client$fn$reify__13620
                                                 onComplete
                                                 "http_client.clj"
                                                 233]
                                                [org.eclipse.jetty.client.ResponseNotifier
                                                 notifyComplete
                                                 "ResponseNotifier.java"
                                                 196]
                                                [org.eclipse.jetty.client.ResponseNotifier
                                                 notifyComplete
                                                 "ResponseNotifier.java"
                                                 188]
                                                [org.eclipse.jetty.client.HttpReceiver
                                                 terminateResponse
                                                 "HttpReceiver.java"
                                                 470]
                                                [org.eclipse.jetty.client.HttpReceiver
                                                 responseSuccess
                                                 "HttpReceiver.java"
                                                 416]
                                                [org.eclipse.jetty.client.http.HttpReceiverOverHTTP
                                                 messageComplete
                                                 "HttpReceiverOverHTTP.java"
                                                 316]
                                                [org.eclipse.jetty.http.HttpParser
                                                 handleContentMessage
                                                 "HttpParser.java"
                                                 599]
                                                [org.eclipse.jetty.http.HttpParser parseNext "HttpParser.java" 1513]
                                                [org.eclipse.jetty.client.http.HttpReceiverOverHTTP
                                                 parse
                                                 "HttpReceiverOverHTTP.java"
                                                 172]
                                                [org.eclipse.jetty.client.http.HttpReceiverOverHTTP
                                                 process
                                                 "HttpReceiverOverHTTP.java"
                                                 135]
                                                [org.eclipse.jetty.client.http.HttpReceiverOverHTTP
                                                 receive
                                                 "HttpReceiverOverHTTP.java"
                                                 73]
                                                [org.eclipse.jetty.client.http.HttpChannelOverHTTP
                                                 receive
                                                 "HttpChannelOverHTTP.java"
                                                 133]
                                                [org.eclipse.jetty.client.http.HttpConnectionOverHTTP
                                                 onFillable
                                                 "HttpConnectionOverHTTP.java"
                                                 155]
                                                [org.eclipse.jetty.io.AbstractConnection$ReadCallback
                                                 succeeded
                                                 "AbstractConnection.java"
                                                 305]
                                                [org.eclipse.jetty.io.FillInterest fillable "FillInterest.java" 103]
                                                [org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint
                                                 onFillable
                                                 "SslConnection.java"
                                                 427]
                                                [org.eclipse.jetty.io.ssl.SslConnection
                                                 onFillable
                                                 "SslConnection.java"
                                                 321]
                                                [org.eclipse.jetty.io.ssl.SslConnection$2
                                                 succeeded
                                                 "SslConnection.java"
                                                 159]
                                                [org.eclipse.jetty.io.FillInterest fillable "FillInterest.java" 103]
                                                [org.eclipse.jetty.io.ChannelEndPoint$2 run "ChannelEndPoint.java" 117]
                                                [org.eclipse.jetty.util.thread.strategy.EatWhatYouKill
                                                 runTask
                                                 "EatWhatYouKill.java"
                                                 333]
                                                [org.eclipse.jetty.util.thread.strategy.EatWhatYouKill
                                                 doProduce
                                                 "EatWhatYouKill.java"
                                                 310]
                                                [org.eclipse.jetty.util.thread.strategy.EatWhatYouKill
                                                 tryProduce
                                                 "EatWhatYouKill.java"
                                                 168]
                                                [org.eclipse.jetty.util.thread.strategy.EatWhatYouKill
                                                 run
                                                 "EatWhatYouKill.java"
                                                 126]
                                                [org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread
                                                 run
                                                 "ReservedThreadExecutor.java"
                                                 366]
                                                [org.eclipse.jetty.util.thread.QueuedThreadPool
                                                 runJob
                                                 "QueuedThreadPool.java"
                                                 765]
                                                [org.eclipse.jetty.util.thread.QueuedThreadPool$2
                                                 run
                                                 "QueuedThreadPool.java"
                                                 683]
                                                [java.lang.Thread run "Thread.java" 834]]}}
```

Here is what the meta for the result looks like:
```
(meta *1)
=>
{:http-request {:request-method :head,
                :scheme :https,
                :server-port 443,
                :uri "REMOVED",
                :headers {"x-amz-date" "20191005T154032Z",
                          "host" "s3.eu-central-1.amazonaws.com",
                          "x-amz-content-sha256" "REMOVED",
                          "authorization" "REMOVED"},
                :body nil,
                :server-name "s3.eu-central-1.amazonaws.com"},
 :http-response {:body nil}}
```

**Comment by jsyrjala** [comment]
Here are the offending line in `http_client.clj`. It seems that it tries to reserve array of 15,395,867,624 bytes, which fails because java max size for array is 2,147,483,647.

```
 34 (defn empty-bbuf
 35   "Returns an array-backed bbuf with pos and lim 0, cap n."
 36   [n]
 37   (.flip (ByteBuffer/wrap (byte-array n))))
```

```
108 (defn- on-headers
109   "Helper for submit. Adds :status, :headers, and :body to state based on
110    response."
111   [state ^Response response]
112   (let [jetty-headers (.getHeaders response)
113         headers (reduce
114                   (fn [m ^String n]
115                     (assoc m
116                       (.toLowerCase n java.util.Locale/ENGLISH)
117                       (.getValue (.getField jetty-headers n))))
118                   {}
119                   (.getFieldNamesCollection jetty-headers))]
120     (assoc state
121       :status (.getStatus response)
122       :headers headers 
123       :body (let [lstr (get headers "content-length")
124                   ln (try
125                       (Long/parseLong lstr)
126                       (catch NumberFormatException _ nil))]
127               (when (and ln (not (zero? ln)))
128                 (empty-bbuf ln))))))
```

**Comment by dchelimsky** [comment]
Next release of aws-api will include an upgrade to cognitect/http-client 0.1.101, which fixes this issue. You can add that explicit dep to fix your project right away, or wait until our next release (soon).

**Comment by dchelimsky** [comment]
Alternatively, if you're using tools.deps, you can change the aws-api dep to this:

``` clojure
com.cognitect.aws/api {:git/url "https://github.com/cognitect-labs/aws-api.git"
                       :sha     "bc44a3ae85f5c450ab9555a7676cac8f6893e5ef"}
```

**Comment by jsyrjala** [comment]
According to [CHANGES.md](https://github.com/cognitect-labs/aws-api/blob/master/CHANGES.md#08391--2019-10-25) this got fixed in 0.8.391. Should this issue be in closed state?

**Comment by dchelimsky** [comment]
Yep!
